### ISSUE TITLE
RAILS_ENV not respected using bin/rake test

### ISSUE THREAD

**Issue by frankpinto** [rails/rails#21478]
Rails version: 4.2.4

**tl;dr** Rails ignores RAILS_ENV and sets the environment to 'test' when running tests

I run `RAILS_ENV="test-kender" bin/rake test test/models/company_test.rb` for a test environment specific to my machine and get the following error:

```
PG::ConnectionBad: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.5432"?
```

I dropped a debugger statement in the initialize method of class PostgreSQL adapter (.../gems/activerecord-4.2.4/lib/active_record/connection_adapters/postgresql_adapter.rb) to inspect connection_parameters and saw it was using the credentials for the 'test' environment. I checked ENV['RAILS_ENV'] and Rails.env and they were, indeed, both set to 'test'

Use case (i.e. why would you want to do that?): http://stackoverflow.com/questions/2861300/can-rails-unit-tests-be-run-on-a-different-environment-than-test

**Comment by kirs** [comment]
Could you share you `test_helper.rb`, please?

**Comment by rafaelfranca** [comment]
Could you try with `DISABLE_SPRING=true RAILS_ENV="test-kender" bin/rake test test/models/company_test.rb`?

**Comment by frankpinto** [comment]
test_helper.rb: https://gist.github.com/frankpinto/789816c049ea535e12df
test: https://gist.github.com/frankpinto/652d358a0721af2fa42d
new command: `DISABLE_SPRING=true RAILS_ENV="test-kender" bin/rake test test/models/company_test.rb test_save_logo`

@rafaelfranca I tried it and got the same error (couldn't connect to Postgres). Here's some more information:

* I checked with the debugger (pry + byebug) again (placed as the first line in PostgreSQLAdapter.initialize). Execution his this line 3 times
* Every time execution hit the debugger I checked Rails.env and it was 'test-kender' everytime
* I also checked the value of 'config' parameter. The first time it corresponded to the config of 'test-kender', the second time it looks like the default config (after this it throws the postgres error), the third time it corresponded to the config of 'test-kender' again.
* The test threw an error at line 14 with the error `NoMethodError: undefined method `logo' for nil:NilClass`

**Comment by matthewd** [comment]
I think this is by design, isn't it?

We *really* don't want to accidentally run the tests against `development` or `production` just because someone has `RAILS_ENV` set.

**Comment by agis-** [comment]
@matthewd If so, we could add a warning when tests are run and `RAILS_ENV` is set by the user to something other than `test`.

What do you think? /cc @rafaelfranca 

P.S. I'd be glad to open a PR for this.

**Comment by rafaelfranca** [comment]
@matthewd yeah make sense.

@agis- a warning is something that could help. Feel free to open a PR.

I'm closing this since it is not an issue.

**Comment by frankpinto** [comment]
The first line in test_helper.rb, outputted by default on new rails projects, is:

    ENV['RAILS_ENV'] ||= 'test'

Given that I really don't think its by design ... that line would be redudant if it were. 

I'm all for smart defaults but they should be overridable. I'll leave it to your discretion as I don't really have time to get a Rails dev install on my machine and work on a PR but I hope its on your radar now just in case anybody else sees this bug. Thanks :)!

**Comment by frankpinto** [comment]
Oh, quick question. What do you all recommend I do for this? Currently whenever I run tests I go into config/database.yml and change the 'test' environment block to my local machines config because of this bug then, when I'm done running tests, I change it back to how it was before to not mess up the other devs on my time. This gets pretty tedious after a bit, any suggestions?

**Comment by rafaelfranca** [comment]
You could set DATABASE_URL if you only want to change the database configuration.

**Comment by matthewd** [comment]
How about this?

```yaml
---
test-kender: &test-kender
  database: db_for_tests
  [..]
test-bob: &test-bob
  database: bobs_test_db
  [..]
test:
  <<: *<%= ENV.fetch('TEST_CONFIG') %>
```

```shell
TEST_CONFIG="test-kender" bin/rake test test/models/company_test.rb
```

**Comment by kirs** [comment]
@rafaelfranca two things I found after reading the issue:
1) actually we allow to pass the environment to the new test runner: https://github.com/rails/rails/blob/master/railties/lib/rails/test_unit/minitest_plugin.rb#L19
2) counting [this line](https://github.com/rails/rails/blob/master/railties/lib/rails/test_unit/minitest_plugin.rb#L42), it makes no sense to add `ENV['RAILS_ENV'] ||= 'test'` to `test_helper.rb` generated in new projects.

what do you thing?

**Comment by rafaelfranca** [comment]
@kirs you are only seeing the Rails test executor but that need to also work with Rake. We need to check if that still make sense for `rake test` too.

**Comment by kirs** [comment]
@rafaelfranca we're [loading the same](https://github.com/rails/rails/blob/master/railties/lib/rails/test_unit/testing.rake#L3) `test_unit/minitest_plugin.rb` in case of Rake

**Comment by matthewd** [comment]
@kirs neither of those helps when running a single test file directly
